pkg/apis/migration/v1alpha1/directimagestreammigration_types.go

```diff
@@ -0,0 +1,128 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"errors"
+
+	imagev1 "github.com/openshift/api/image/v1"
+	kapi "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DirectImageStreamMigrationSpec defines the desired state of DirectImageStreamMigration
+type DirectImageStreamMigrationSpec struct {
+	SrcMigClusterRef  *kapi.ObjectReference `json:"srcMigClusterRef,omitempty"`
+	DestMigClusterRef *kapi.ObjectReference `json:"destMigClusterRef,omitempty"`
+	ImageStreamRef    *kapi.ObjectReference `json:"imageStreamRef,omitempty"`
+	DestNamespace     string                `json:"destNamespace,omitempty"`
+}
+
+// DirectImageStreamMigrationStatus defines the observed state of DirectImageStreamMigration
+type DirectImageStreamMigrationStatus struct {
+	Conditions
+	ObservedDigest string       `json:"observedDigest,omitempty"`
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+	Phase          string       `json:"phase,omitempty"`
+	Itinerary      string       `json:"itinerary,omitempty"`
+	Errors         []string     `json:"errors,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DirectImageStreamMigration is the Schema for the directimagestreammigrations API
+// +kubebuilder:resource:path=directimagestreammigrations,shortName=dism
+// +k8s:openapi-gen=true
+type DirectImageStreamMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   DirectImageStreamMigrationSpec   `json:"spec,omitempty"`
+	Status DirectImageStreamMigrationStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DirectImageStreamMigrationList contains a list of DirectImageStreamMigration
+type DirectImageStreamMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DirectImageStreamMigration `json:"items"`
+}
+
+func (r *DirectImageStreamMigration) GetSourceCluster(client k8sclient.Client) (*MigCluster, error) {
+	return GetCluster(client, r.Spec.SrcMigClusterRef)
+}
+
+func (r *DirectImageStreamMigration) GetDestinationCluster(client k8sclient.Client) (*MigCluster, error) {
+	return GetCluster(client, r.Spec.DestMigClusterRef)
+}
+
+func (r *DirectImageStreamMigration) GetImageStream(c k8sclient.Client) (*imagev1.ImageStream, error) {
+	cluster, err := r.GetSourceCluster(c)
+	if err != nil {
+		return nil, err
+	}
+	if !cluster.Status.IsReady() {
+		return nil, errors.New("Source cluster not ready")
```

Comment:
Probably worth defining the error type for "Source cluster not ready" could see how it can be used at a lot of places in mig-controller code